File: app/api/v1/preorders.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Path, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.db import get_async_db
from app.core.dependencies import (
    Pagination,
    api_rate_limit,
    get_current_verified_user,
    get_pagination,
    require_company_access,
    require_store_roles,
)
from app.core.security import resolve_tenant_company_id
from app.models.user import User
from app.schemas.preorders import (
    PreorderCreateIn,
    PreorderListFilters,
    PreorderListResponse,
    PreorderOut,
    PreorderUpdateIn,
)
from app.services.preorders import (
    cancel_preorder,
    confirm_preorder,
    create_preorder,
    fulfill_preorder,
    get_preorder,
    list_preorders,
    update_preorder,
)

router = APIRouter()


async def _require_company_context(current_user: User = Depends(get_current_verified_user)) -> User:
    resolve_tenant_company_id(current_user, not_found_detail="Company not set")
    return current_user


store_router = APIRouter(
    prefix="/preorders",
    tags=["preorders"],
    dependencies=[
        Depends(api_rate_limit),
        Depends(require_company_access),
        Depends(require_store_roles("admin")),
        Depends(_require_company_context),
    ],
)


@store_router.post("", response_model=PreorderOut, status_code=status.HTTP_201_CREATED)
async def create_preorder_endpoint(
    payload: PreorderCreateIn,
    current_user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_async_db),
):
    company_id = resolve_tenant_company_id(current_user, not_found_detail="Company not set")
    preorder = await create_preorder(
        db,
        company_id=company_id,
        created_by_user_id=getattr(current_user, "id", None),
        payload=payload,
    )
    return PreorderOut.model_validate(preorder)


@store_router.get("", response_model=PreorderListResponse)
async def list_preorders_endpoint(
    filters: PreorderListFilters = Depends(),
    pagination: Pagination = Depends(get_pagination),
    current_user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_async_db),
):
    company_id = resolve_tenant_company_id(current_user, not_found_detail="Company not set")
    items, total = await list_preorders(
        db,
        company_id=company_id,
        filters=filters,
        offset=pagination.offset,
        limit=pagination.limit,
    )
    payload_items = [PreorderOut.model_validate(item) for item in items]
    return PreorderListResponse.create(
        items=payload_items, total=total, page=pagination.page, per_page=pagination.per_page
    )


@store_router.get("/{preorder_id}", response_model=PreorderOut)
async def get_preorder_endpoint(
    preorder_id: int = Path(..., ge=1),
    current_user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_async_db),
):
    company_id = resolve_tenant_company_id(current_user, not_found_detail="Company not set")
    preorder = await get_preorder(db, company_id=company_id, preorder_id=preorder_id)
    return PreorderOut.model_validate(preorder)


@store_router.patch("/{preorder_id}", response_model=PreorderOut)
async def update_preorder_endpoint(
    payload: PreorderUpdateIn,
    preorder_id: int = Path(..., ge=1),
    current_user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_async_db),
):
    company_id = resolve_tenant_company_id(current_user, not_found_detail="Company not set")
    preorder = await update_preorder(db, company_id=company_id, preorder_id=preorder_id, payload=payload)
    return PreorderOut.model_validate(preorder)


@store_router.post("/{preorder_id}/confirm", response_model=PreorderOut)
async def confirm_preorder_endpoint(
    preorder_id: int = Path(..., ge=1),
    current_user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_async_db),
):
    company_id = resolve_tenant_company_id(current_user, not_found_detail="Company not set")
    preorder = await confirm_preorder(db, company_id=company_id, preorder_id=preorder_id)
    return PreorderOut.model_validate(preorder)


@store_router.post("/{preorder_id}/cancel", response_model=PreorderOut)
async def cancel_preorder_endpoint(
    preorder_id: int = Path(..., ge=1),
    current_user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_async_db),
):
    company_id = resolve_tenant_company_id(current_user, not_found_detail="Company not set")
    preorder = await cancel_preorder(db, company_id=company_id, preorder_id=preorder_id)
    return PreorderOut.model_validate(preorder)


@store_router.post("/{preorder_id}/fulfill", response_model=PreorderOut)
async def fulfill_preorder_endpoint(
    preorder_id: int = Path(..., ge=1),
    current_user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_async_db),
):
    company_id = resolve_tenant_company_id(current_user, not_found_detail="Company not set")
    preorder = await fulfill_preorder(db, company_id=company_id, preorder_id=preorder_id)
    return PreorderOut.model_validate(preorder)


router.include_router(store_router)

File: frontend/src/api/preorders.ts
import { apiClient } from './client'

export type PaginatedResponse<T> = {
  items: T[]
  total: number
  page: number
  per_page: number
  pages: number
  has_next: boolean
  has_prev: boolean
}

export type PreorderItemIn = {
  product_id?: number | null
  sku?: string | null
  name?: string | null
  qty: number
  price?: number | string | null
}

export type PreorderCreateIn = {
  currency?: string
  customer_name?: string | null
  customer_phone?: string | null
  notes?: string | null
  items?: PreorderItemIn[]
}

export type PreorderUpdateIn = {
  customer_name?: string | null
  customer_phone?: string | null
  notes?: string | null
  items?: PreorderItemIn[] | null
}

export type PreorderItemOut = {
  id: number
  created_at: string
  updated_at: string
  product_id?: number | null
  sku?: string | null
  name?: string | null
  qty: number
  price?: string | null
  preorder_id: number
}

export type Preorder = {
  id: number
  created_at: string
  updated_at: string
  company_id: number
  status: string
  currency: string
  total: string | null
  customer_name: string | null
  customer_phone: string | null
  notes: string | null
  created_by_user_id: number | null
  fulfilled_order_id: number | null
  fulfilled_at: string | null
  items?: PreorderItemOut[] | null
}

export type PreorderListParams = {
  status?: string | null
  date_from?: string | null
  date_to?: string | null
  page?: number
  per_page?: number
}

export async function listPreorders(
  params: PreorderListParams = {}
): Promise<PaginatedResponse<Preorder>> {
  const { data } = await apiClient.get<PaginatedResponse<Preorder>>('/api/v1/preorders', { params })
  return data
}

export async function getPreorderById(preorderId: number): Promise<Preorder> {
  const { data } = await apiClient.get<Preorder>(`/api/v1/preorders/${preorderId}`)
  return data
}

export async function createPreorder(payload: PreorderCreateIn): Promise<Preorder> {
  const { data } = await apiClient.post<Preorder>('/api/v1/preorders', payload)
  return data
}

export async function updatePreorder(preorderId: number, payload: PreorderUpdateIn): Promise<Preorder> {
  const { data } = await apiClient.patch<Preorder>(`/api/v1/preorders/${preorderId}`, payload)
  return data
}

export async function confirmPreorder(preorderId: number): Promise<Preorder> {
  const { data } = await apiClient.post<Preorder>(`/api/v1/preorders/${preorderId}/confirm`)
  return data
}

export async function cancelPreorder(preorderId: number): Promise<Preorder> {
  const { data } = await apiClient.post<Preorder>(`/api/v1/preorders/${preorderId}/cancel`)
  return data
}

export async function fulfillPreorder(preorderId: number): Promise<Preorder> {
  const { data } = await apiClient.post<Preorder>(`/api/v1/preorders/${preorderId}/fulfill`)
  return data
}
